bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/ResumingStreamingResultScanner.java

```diff
@@ -94,27 +95,48 @@ public Row next() throws IOException {
         if (result != null) {
           lastRowKey = result.getKey();
           rowCount++;
-          // We've had at least one successful RPC, reset the backoff
-          currentBackoff = null;
+          // We've had at least one successful RPC, reset the backoffs
+          currentErrorBackoff = null;
+          currentTimeoutBackoff = null;
         }
 
         return result;
       } catch (ScanTimeoutException rte) {
-        logger.info("The client could not get a response in %d ms. Retrying the scan.",
-          retryOptions.getReadPartialRowTimeoutMillis());
-        backOffAndRetry(rte);
+        handleScanTimeout(rte);
       } catch (IOExceptionWithStatus ioe) {
-        Status.Code code = ioe.getStatus().getCode();
-        if (retryOptions.isRetryable(code)) {
-          logger.info("Reissuing scan after receiving error with status: %s.", code.name());
-          backOffAndRetry(ioe);
-        } else {
-          throw ioe;
-        }
+        handleIOException(ioe);
       }
     }
   }
 
+  private void handleScanTimeout(ScanTimeoutException rte) throws IOException {
+    logger.info("The client could not get a response in %d ms. Retrying the scan.",
+      retryOptions.getReadPartialRowTimeoutMillis());
+    if (currentTimeoutBackoff == null) {
+      // Create a backoff that retries immediately, up to the max number of times
+      currentTimeoutBackoff =
+          new TimeoutRetryBackOff(retryOptions.getMaxScanTimeoutRetries(), BackOff.ZERO_BACKOFF);
+    }
+    // Reset the error backoff in case we encountered this timeout after an error.
+    // Otherwise, we will likely have already exceeded the max elapsed time for the backoff
+    // and won't retry after the next error.
+    currentErrorBackoff = null;
+    backOffAndRetry(currentTimeoutBackoff, rte);
+  }
+
+  private void handleIOException(IOExceptionWithStatus ioe) throws IOException {
```

Comment:
If we reset currentTimeoutBackoff here then we don't have a mechanism to ever give up if we, for example, alternate between timeouts and retryable exceptions. By not resetting here we're saying that we will never accept more than a fixed number of timeouts for a given request which seems appropriate.